### ISSUE TITLE
Tensorflow C++ error when using CV Mat image 

### ISSUE THREAD

**Issue by shashank2710** [tensorflow/tensorflow#22068]
I am trying to create a custom OCR engine using Tensorflow(v1.10) in Ubuntu 18.04, with Bazel (0.16) in C++.
------------------------
## Issues Faced
- Core Dumped Error for dtype mismatch (0 vs 1)
- Predicted value of weights is higher than 1, when Tensorflow Session is re-run in loop
-----------------------
#### Code

```
for (size_t i=0; i<filenames.size(); i++)
	{
		cv::Mat image = cv::imread(filenames[i]);

		Tensor image_tensor (tensorflow::DT_FLOAT,tensorflow::TensorShape{1,tf_height,tf_width,3});
		image.convertTo(image, CV_32FC1);
		tensorflow::StringPiece tmp_data = image_tensor.tensor_data();
		memcpy(const_cast<char*>(tmp_data.data()), (image.data), tf_height * tf_width * sizeof(float));

		// Creating a Session with the Graph
std::unique_ptr<tensorflow::Session>session(tensorflow::NewSession(tensorflow::SessionOptions()));
		//session->tensorflow::reset(tensorflow::NewSession(tensorflow::SessionOptions()));
		tensorflow::Status session_create_status = session->Create(graph_def);

		std::vector<std::pair<string, tensorflow::Tensor>> inputs = {{inputLayer, image_tensor}};
		outputs.clear();
		Status runStatus = session->Run(inputs, {outputLayer}, {}, &outputs);
		if (!runStatus.ok()) {
			LOG(ERROR) << "Running model failed: " << runStatus;
			return -1;
		}
		cv::imshow("Original Image",image);
		cv::waitKey();
		session->Close();
		image.release();
	}
```

-----------------------

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by PatWie** [comment]
That `CV_32FC1` is supposed to be a [`CV_32FC3`](https://github.com/PatWie/tensorflow-cmake/blob/master/examples/resize/opencv_version.cc#L30) since your tensor should have 3 channels, right?
I am pretty sure SO would be the better place to ask this question.

**Comment by shashank2710** [comment]
Have I written custom code: yes
OS Platform and Distribution: Ubuntu 18.04
TensorFlow installed from: Source
TensorFlow version: 1.10 (CPU version)
Bazel version:0.16
CUDA/cuDNN version: 9.0 (installed but not used)
GPU model and memory: Nvidia 1070 (present but not used)
Exact command to reproduce: N/A
Mobile device No

**Comment by shashank2710** [comment]
@PatWie.. Thank you for your response. That was helpful.  

Your link that you posted is not valid. Could you kindly update that? 

@PatWie @skye 
Attached herewith is an updated code that I am currently working on:
```
/*
 * main.cpp
 *
 *  Created on: Aug 21, 2018
 *      Author: ssatyanarayana
 */


#include <fstream>
#include <utility>
#include <vector>
#include <iostream>

#include "tensorflow/cc/ops/const_op.h"
#include "tensorflow/cc/ops/image_ops.h"
#include "tensorflow/cc/ops/standard_ops.h"
#include "tensorflow/core/framework/graph.pb.h"
#include "tensorflow/core/graph/default_device.h"
#include "tensorflow/core/graph/graph_def_builder.h"
#include "tensorflow/core/lib/core/threadpool.h"
#include "tensorflow/core/lib/io/path.h"
#include "tensorflow/core/lib/strings/stringprintf.h"
#include "tensorflow/core/platform/init_main.h"
#include "tensorflow/core/public/session.h"
#include "tensorflow/core/util/command_line_flags.h"

#include <opencv2/core/mat.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>

using tensorflow::Flag;
using tensorflow::Tensor;
using tensorflow::Status;
using tensorflow::string;
using tensorflow::int32;

using namespace std;
using namespace cv;


const int64 tf_height = 128;
const int64 tf_width = 128;

int main()
{
	string folderpath = "/home/ssatyanarayana/OCR-Engine-Local/data/alphabets/myTest/*.jpg";
	std::vector<String> filenames;
	cv::glob(folderpath, filenames);

	string LABELS = "/home/ssatyanarayana/OCR-Engine-Local/k2tf-20180827143102/indices.txt";

	// Set input & output nodes names
	std::string inputLayer = "conv2d_1_input";
	std::string outputLayer = "k2tfout_0";

	//Initializing the Graph
	tensorflow::GraphDef graph_def;

	// Specify file location of Output Graph
	std::string graphFile = "/home/ssatyanarayana/OCR-Engine-Local/output_graph.pb";

	// Loading the graph to the given variable
	tensorflow::Status graphLoadedStatus = ReadBinaryProto(tensorflow::Env::Default(),graphFile,&graph_def);
	if (!graphLoadedStatus.ok()){
		std::cout << graphLoadedStatus.ToString()<<std::endl;
		return 1;
	}

	std::vector<Tensor> outputs;

	for (size_t i=0; i<filenames.size(); i++)
	{
		cv::Mat image = cv::imread(filenames[i]);

		Tensor image_tensor (tensorflow::DT_FLOAT, tensorflow::TensorShape{1,tf_height,tf_width,3});
		image.convertTo(image, CV_32FC3);
		tensorflow::StringPiece tmp_data = image_tensor.tensor_data();
		memcpy(const_cast<char*>(tmp_data.data()), (image.data), tf_height * tf_width * sizeof(float));

		// Creating a Session with the Graph
		std::unique_ptr<tensorflow::Session> session(tensorflow::NewSession(tensorflow::SessionOptions()));
		//session->tensorflow::reset(tensorflow::NewSession(tensorflow::SessionOptions()));
		tensorflow::Status session_create_status = session->Create(graph_def);

		std::vector<std::pair<string, tensorflow::Tensor>> inputs = {{inputLayer, image_tensor}};
		outputs.clear();

		Status runStatus = session->Run(inputs, {outputLayer}, {}, &outputs);
		if (!runStatus.ok()) {
			LOG(ERROR) << "Running model failed: " << runStatus;
			return -1;
		}

		for (auto &t : outputs)
		{
			std::cout << t.DebugString()<<std::endl;
			tensorflow::TTypes<float, 2>::Tensor scores = t.flat_inner_dims<float>();
			auto dims = scores.dimensions();
			int imgCount = dims[0];
			int classesCount = dims[1];
			for(int i = 0; i<imgCount; i++)
			{
				float maxVal = scores(i,0);
				int maxIndex = 0;
				for(int j = 1; j<classesCount; j++) {
					float val = scores(i,j);
					if(val > maxVal) {
						maxVal = val;
						maxIndex = j;
					}
				}
				std::cout << "Img" << to_string(i) << " prediction: " << to_string(maxIndex) << ", score: " << to_string(maxVal)<<std::endl;
			}
		}

		cv::imshow("Original Image",image);
		cv::waitKey();
		session->Close();
		image.release();
	}
    return 0;
}


```

## Issues Faced:
-Output tensor fails to clear and provide accurate weights for consecutive loops. 
-Same graph provides higher accuracy in Python (with Keras as TF as backend) when compared to using in C++ directly

You can access all the required files on [https://github.com/shashank2710/OCR-Engine](url)

**Comment by skye** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!
